README
NAME

    File::Codeowners - Read and write CODEOWNERS files

VERSION

    version 0.52

METHODS

 new

        $codeowners = File::Codeowners->new;

    Construct a new File::Codeowners.

 parse

        $codeowners = File::Codeowners->parse('path/to/CODEOWNERS');
        $codeowners = File::Codeowners->parse($filehandle);
        $codeowners = File::Codeowners->parse(\@lines);
        $codeowners = File::Codeowners->parse(\$string);

    Parse a CODEOWNERS file.

    This is a shortcut for the parse_from_* methods.

 parse_from_filepath

        $codeowners = File::Codeowners->parse_from_filepath('path/to/CODEOWNERS');

    Parse a CODEOWNERS file from the filesystem.

 parse_from_fh

        $codeowners = File::Codeowners->parse_from_fh($filehandle);

    Parse a CODEOWNERS file from an open filehandle.

 parse_from_array

        $codeowners = File::Codeowners->parse_from_array(\@lines);

    Parse a CODEOWNERS file stored as lines in an array.

 parse_from_string

        $codeowners = File::Codeowners->parse_from_string(\$string);
        $codeowners = File::Codeowners->parse_from_string($string);

    Parse a CODEOWNERS file stored as a string. String should be UTF-8
    encoded.

 write_to_filepath

        $codeowners->write_to_filepath($filepath);

    Write the contents of the file to the filesystem atomically.

 write_to_fh

        $codeowners->write_to_fh($fh);

    Format the file contents and write to a filehandle.

 write_to_string

        $scalarref = $codeowners->write_to_string;

    Format the file contents and return a reference to a formatted string.

 write_to_array

        $lines = $codeowners->write_to_array;

    Format the file contents as an arrayref of lines.

 match

        $owners = $codeowners->match($filepath);

    Match the given filepath against the available patterns and return just
    the owners for the matching pattern. Patterns are checked in the
    reverse order they were defined in the file.

    Returns undef if no patterns match.

 owners

        $owners = $codeowners->owners; # get all defined owners
        $owners = $codeowners->owners($pattern);

    Get an arrayref of owners defined in the file. If a pattern argument is
    given, only owners for the given pattern are returned (or empty
    arrayref if the pattern does not exist). If no argument is given,
    simply returns all owners defined in the file.

 patterns

        $patterns = $codeowners->patterns;
        $patterns = $codeowners->patterns($owner);

    Get an arrayref of all patterns defined.

 projects

        $projects = $codeowners->projects;

    Get an arrayref of all projects defined.

 update_owners

        $codeowners->update_owners($pattern => \@new_owners);

    Set a new set of owners for a given pattern. If for some reason the
    file has multiple such patterns, they will all be updated.

    Nothing happens if the file does not already have at least one such
    pattern.

 update_owners_by_project

        $codeowners->update_owners_by_project($project => \@new_owners);

    Set a new set of owners for all patterns under the given project.

    Nothing happens if the file does not have a project with the given
    name.

 rename_owner

        $codeowners->rename_owner($old_name => $new_name);

    Rename an owner.

    Nothing happens if the file does not have an owner with the old name.

 rename_project

        $codeowners->rename_project($old_name => $new_name);

    Rename a project.

    Nothing happens if the file does not have a project with the old name.

 append

        $codeowners->append(comment => $str);
        $codeowners->append(pattern => $pattern, owners => \@owners);
        $codeowners->append();     # blank line

    Append a new line.

 prepend

        $codeowners->prepend(comment => $str);
        $codeowners->prepend(pattern => $pattern, owners => \@owners);
        $codeowners->prepend();    # blank line

    Prepend a new line.

 unowned

        $filepaths = $codeowners->unowned;

    Get the list of filepaths in the "unowned" section.

    This parser supports an "extension" to the CODEOWNERS file format which
    lists unowned files at the end of the file. This list can be useful to
    have in order to figure out what files we know are unowned versus what
    files we don't know are unowned.

 add_unowned

        $codeowners->add_unowned($filepath, ...);

    Add one or more filepaths to the "unowned" list.

    This method does not check to make sure the filepath(s) actually do not
    match any patterns in the file, so you might want to call "match"
    first.

    See "unowned" for an explanation.

 remove_unowned

        $codeowners->remove_unowned($filepath, ...);

    Remove one or more filepaths from the "unowned" list.

    Silently ignores filepaths that are already not listed.

    See "unowned" for an explanation.

 is_unowned

        $bool = $codeowners->is_unowned($filepath);

    Test whether a filepath is in the "unowned" list.

    See "unowned" for an explanation.

 clear_unowned

        $codeowners->clear_unowned;

    Remove all filepaths from the "unowned" list.

    See "unowned" for an explanation.

BUGS

    Please report any bugs or feature requests on the bugtracker website
    https://github.com/chazmcgarvey/File-Codeowners/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR

    Charles McGarvey <chazmcgarvey@brokenzipper.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2021 by Charles McGarvey.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for File-Codeowners.

{{$NEXT}}

0.55      2022-06-29 21:50:55-0600
  * Fix flappy unit test on Win32.

0.54      2022-06-28 16:24:46-06:00 America/Denver
  * Add support for group alias defined in CODEOWNERS files.

0.53      2021-04-02 19:47:45-06:00 America/Denver
  * Test::File::Codeowners now looks for CODEOWNERS files and repos
    relative to the executing test script (not cwd).

0.52      2021-04-02 19:07:53-06:00 America/Denver
  * Handle no-git and no-repo edge cases in Test::File::Codeowners.

0.51      2021-04-02 14:59:55-06:00 America/Denver
  * Add Test::File::Codeowners from App-Codeowners.

0.50      2021-04-02 14:06:24-06:00 America/Denver
  * Split from App-Codeowners.


lib/Test/File/Codeowners.pm
package Test::File::Codeowners;
# ABSTRACT: Write tests for CODEOWNERS files

=head1 SYNOPSIS

    use Test::More;

    eval 'use Test::File::Codeowners';
    plan skip_all => 'Test::File::Codeowners required for testing CODEOWNERS' if $@;

    codeowners_syntax_ok();
    done_testing;

=head1 DESCRIPTION

This package has assertion subroutines for testing F<CODEOWNERS> files.

=cut

use warnings;
use strict;

use Encode qw(encode);
use File::Codeowners::Util qw(find_codeowners_in_directory find_nearest_codeowners git_ls_files git_toplevel);
use File::Codeowners;
use FindBin qw($Bin);
use Test::Builder;

our $VERSION = '9999.999'; # VERSION

my $Test = Test::Builder->new;

sub import {
    my $self = shift;
    my $caller = caller;
    no strict 'refs';   ## no critic (TestingAndDebugging::ProhibitNoStrict)
    *{$caller.'::codeowners_syntax_ok'} = \&codeowners_syntax_ok;
    *{$caller.'::codeowners_git_files_ok'} = \&codeowners_git_files_ok;

    $Test->exported_to($caller);
    $Test->plan(@_);
}

=func codeowners_syntax_ok

    codeowners_syntax_ok();     # search up the tree for a CODEOWNERS file
    codeowners_syntax_ok($filepath);

Check the syntax of a F<CODEOWNERS> file.

=cut

sub codeowners_syntax_ok {
    my $filepath = shift || find_nearest_codeowners($Bin);

    if (!$filepath) {
        $Test->ok(0, "Check syntax: <missing>");
        $Test->diag('No CODEOWNERS file could be found.');
        return;
    }

    eval { File::Codeowners->parse($filepath) };
    my $err = $@;

    $Test->ok(!$err, "Check syntax: $filepath");
    $Test->diag($err) if $err;
}

=func codeowners_git_files_ok

    codeowners_git_files_ok();  # use git repo in cwd
    codeowners_git_files_ok($repopath);

=cut

sub codeowners_git_files_ok {
    my $repopath = shift || $Bin;

    my $git_toplevel = git_toplevel($repopath);
    if (!$git_toplevel) {
        $Test->skip('No git repo could be found.');
        return;
    }

    my $filepath = find_codeowners_in_directory($git_toplevel);
    if (!$filepath) {
        $Test->ok(0, "Check syntax: <missing>");
        $Test->diag("No CODEOWNERS file could be found in repo $repopath.");
        return;
    }

    $Test->subtest('codeowners_git_files_ok' => sub {
        local $Test::Builder::Level = $Test::Builder::Level + 3;

        my $codeowners = eval { File::Codeowners->parse($filepath) };
        if (my $err = $@) {
            $Test->plan(tests => 1);
            $Test->ok(0, "Parse $filepath");
            $Test->diag($err);
            return;
        }

        my ($proc, @files) = git_ls_files($git_toplevel);
        if ($proc->wait != 0) {
            $Test->plan(skip_all => 'git ls-files failed');
            return;
        }

        $Test->plan(tests => scalar @files);

        for my $filepath (@files) {
            my $msg = encode('UTF-8', "Check file: $filepath");

            my $match = $codeowners->match($filepath);
            my $is_unowned = $codeowners->is_unowned($filepath);

            if (!$match && !$is_unowned) {
                $Test->ok(0, $msg);
                $Test->diag("File is unowned\n");
            }
            elsif ($match && $is_unowned) {
                $Test->ok(0, $msg);
                $Test->diag("File is owned but listed as unowned\n");
            }
            else {
                $Test->ok(1, $msg);
                if ($match) {
                    my $owners = encode('UTF-8', join(',', @{$match->{owners}}));
                    $Test->note("File is owned by $owners");
                }
            }
        }
    });
}

1;


lib/File/Codeowners.pm
package File::Codeowners;
# ABSTRACT: Read and write CODEOWNERS files

use v5.10.1;    # defined-or
use warnings;
use strict;

use Encode qw(encode);
use Path::Tiny 0.089;
use Scalar::Util qw(openhandle);
use Text::Gitignore qw(build_gitignore_matcher);

our $VERSION = '9999.999'; # VERSION

my $RE_PATTERN  = qr/.+?(?<!\\)/;
my $RE_OWNER    = qr/(?:\@+"[^"]*")|(?:\H+)/;

sub _croak { require Carp; Carp::croak(@_); }
sub _usage { _croak("Usage: @_\n") }

=method new

    $codeowners = File::Codeowners->new;

Construct a new L<File::Codeowners>.

=cut

sub new {
    my $class = shift;
    my $self  = bless {}, $class;
}

=method parse

    $codeowners = File::Codeowners->parse($filepath, @options);
    $codeowners = File::Codeowners->parse(*IO, @options);
    $codeowners = File::Codeowners->parse(\@lines, @options);
    $codeowners = File::Codeowners->parse(\$string, @options);

Parse a F<CODEOWNERS> file.

This is a shortcut for the C<parse_from_*> methods.

Possible options:

=for :list
* C<aliases> - Parse lines that begin with "@" as aliases (default: false)

=cut

sub parse {
    my $self  = shift;
    my $input = shift or _usage(q{$codeowners->parse($input)});

    return $self->parse_from_array($input, @_)  if ref($input) eq 'ARRAY';
    return $self->parse_from_string($input, @_) if ref($input) eq 'SCALAR';
    return $self->parse_from_fh($input, @_)     if openhandle($input);
    return $self->parse_from_filepath($input, @_);
}

=method parse_from_filepath

    $codeowners = File::Codeowners->parse_from_filepath($filepath, @options);

Parse a F<CODEOWNERS> file from the filesystem.

=cut

sub parse_from_filepath {
    my $self = shift;
    my $path = shift or _usage(q{$codeowners->parse_from_filepath($filepath)});

    $self = bless({}, $self) if !ref($self);

    return $self->parse_from_fh(path($path)->openr_utf8, @_);
}

=method parse_from_fh

    $codeowners = File::Codeowners->parse_from_fh(*IO, @options);

Parse a F<CODEOWNERS> file from an open filehandle.

=cut

sub parse_from_fh {
    my $self = shift;
    my $fh   = shift or _usage(q{$codeowners->parse_from_fh($fh)});
    my %opts = @_;

    $self = bless({}, $self) if !ref($self);

    my @lines;

    my $parse_unowned;
    my %unowned;
    my %aliases;
    my $current_project;

    while (my $line = <$fh>) {
        my $lineno = $. - 1;
        chomp $line;
        if ($line eq '### UNOWNED (File::Codeowners)') {
            $parse_unowned++;
            last;
        }
        elsif ($line =~ /^\h*#(.*)/) {
            my $comment = $1;
            my $project;
            if ($comment =~ /^\h*Project:\h*(.+?)\h*$/i) {
                $project = $current_project = $1 || undef;
            }
            $lines[$lineno] = {
                comment => $comment,
                $project ? (project => $project) : (),
            };
        }
        elsif ($line =~ /^\h*$/) {
            # blank line
        }
        elsif ($opts{aliases} && $line =~ /^\h*\@($RE_OWNER)\h+(.+)/) {
            my $alias   = $1;
            my @owners  = $2 =~ /($RE_OWNER)/g;
            $aliases{$alias} = \@owners;
            $lines[$lineno] = {
                alias   => $alias,
                owners  => \@owners,
            };
        }
        elsif ($line =~ /^\h*($RE_PATTERN)\h+(.+)/) {
            my $pattern = $1;
            my @owners  = $2 =~ /($RE_OWNER)/g;
            $lines[$lineno] = {
                pattern => $pattern,
                owners  => \@owners,
                $current_project ? (project => $current_project) : (),
            };
        }
        else {
            die "Parse error on line $.: $line\n";
        }
    }

    if ($parse_unowned) {
        while (my $line = <$fh>) {
            chomp $line;
            if ($line =~ /# (.+)/) {
                my $filepath = $1;
                $unowned{$filepath}++;
            }
        }
    }

    $self->{lines} = \@lines;
    $self->{unowned} = \%unowned;
    $self->{aliases} = \%aliases;

    return $self;
}

=method parse_from_array

    $codeowners = File::Codeowners->parse_from_array(\@lines, @options);

Parse a F<CODEOWNERS> file stored as lines in an array.

=cut

sub parse_from_array {
    my $self = shift;
    my $arr  = shift or _usage(q{$codeowners->parse_from_array(\@lines)});

    $self = bless({}, $self) if !ref($self);

    my $str = join("\n", @$arr);
    return $self->parse_from_string(\$str, @_);
}

=method parse_from_string

    $codeowners = File::Codeowners->parse_from_string(\$string, @options);
    $codeowners = File::Codeowners->parse_from_string($string, @options);

Parse a F<CODEOWNERS> file stored as a string. String should be UTF-8 encoded.

=cut

sub parse_from_string {
    my $self = shift;
    my $str  = shift or _usage(q{$codeowners->parse_from_string(\$string)});

    $self = bless({}, $self) if !ref($self);

    my $ref = ref($str) eq 'SCALAR' ? $str : \$str;
    open(my $fh, '<:encoding(UTF-8)', $ref) or die "open failed: $!";

    return $self->parse_from_fh($fh, @_);
}

=method write_to_filepath

    $codeowners->write_to_filepath($filepath);

Write the contents of the file to the filesystem atomically.

=cut

sub write_to_filepath {
    my $self = shift;
    my $path = shift or _usage(q{$codeowners->write_to_filepath($filepath)});

    path($path)->spew_utf8([map { "$_\n" } @{$self->write_to_array}]);
}

=method write_to_fh

    $codeowners->write_to_fh($fh);

Format the file contents and write to a filehandle.

=cut

sub write_to_fh {
    my $self    = shift;
    my $fh      = shift or _usage(q{$codeowners->write_to_fh($fh)});
    my $charset = shift;

    for my $line (@{$self->write_to_array($charset)}) {
        print $fh "$line\n";
    }
}

=method write_to_string

    \$string = $codeowners->write_to_string;

Format the file contents and return a reference to a formatted string.

=cut

sub write_to_string {
    my $self    = shift;
    my $charset = shift;

    my $str = join("\n", @{$self->write_to_array($charset)}) . "\n";
    return \$str;
}

=method write_to_array

    \@lines = $codeowners->write_to_array;

Format the file contents as an arrayref of lines.

=cut

sub write_to_array {
    my $self    = shift;
    my $charset = shift;

    my @format;

    for my $line (@{$self->_lines}) {
        if (my $comment = $line->{comment}) {
            push @format, "#$comment";
        }
        elsif (my $pattern = $line->{pattern}) {
            my $owners = join(' ', @{$line->{owners}});
            push @format, "$pattern  $owners";
        }
        elsif (my $alias = $line->{alias}) {
            my $owners = join(' ', @{$line->{owners}});
            push @format, "\@$alias  $owners";
        }
        else {
            push @format, '';
        }
    }

    my @unowned = sort keys %{$self->_unowned};
    if (@unowned) {
        push @format, '' if $format[-1];
        push @format, '### UNOWNED (File::Codeowners)';
        for my $unowned (@unowned) {
            push @format, "# $unowned";
        }
    }

    if (defined $charset) {
        $_ = encode($charset, $_) for @format;
    }
    return \@format;
}

=method match

    \%match = $codeowners->match($filepath, %options);

Match the given filepath against the available patterns and return just the
owners for the matching pattern. Patterns are checked in the reverse order
they were defined in the file.

Returns C<undef> if no patterns match.

Possible options:

=for :list
* C<expand> - Expand group aliases defined in the F<CODEOWNERS> file.

=cut

sub match {
    my $self     = shift;
    my $filepath = shift or _usage(q{$codeowners->match($filepath)});
    my %opts     = @_;

    my $expand = $opts{expand} ? do {
        my $aliases = $self->aliases;
        sub {
            my $owner = shift;
            my $alias = $aliases->{$owner};
            return @$alias if $alias;
            return $owner;
        };
    } : sub { shift };  # noop

    my $lines = $self->{match_lines} ||= [reverse grep { ($_ || {})->{pattern} } @{$self->_lines}];

    for my $line (@$lines) {
        my $matcher = $line->{matcher} ||= build_gitignore_matcher([$line->{pattern}]);
        return {    # deep copy
            pattern => $line->{pattern},
            owners  => [map { $expand->($_) } @{$line->{owners} || []}],
            $line->{project} ? (project => $line->{project}) : (),
        } if $matcher->($filepath);
    }

    return undef;   ## no critic (Subroutines::ProhibitExplicitReturn)
}

=method owners

    $owners = $codeowners->owners; # get all defined owners
    $owners = $codeowners->owners($pattern);

Get an arrayref of owners defined in the file. If a pattern argument is given,
only owners for the given pattern are returned (or empty arrayref if the
pattern does not exist). If no argument is given, simply returns all owners
defined in the file.

=cut

sub owners {
    my $self    = shift;
    my $pattern = shift;

    return $self->{owners} if !$pattern && $self->{owners};

    my %owners;
    for my $line (@{$self->_lines}) {
        next if $pattern && $line->{pattern} && $pattern ne $line->{pattern};
        $owners{$_}++ for (@{$line->{owners} || []});
    }

    my $owners = [sort keys %owners];
    $self->{owners} = $owners if !$pattern;

    return $owners;
}

=method patterns

    $patterns = $codeowners->patterns;
    $patterns = $codeowners->patterns($owner);

Get an arrayref of all patterns defined.

=cut

sub patterns {
    my $self  = shift;
    my $owner = shift;

    return $self->{patterns} if !$owner && $self->{patterns};

    my %patterns;
    for my $line (@{$self->_lines}) {
        next if $owner && !grep { $_ eq $owner  } @{$line->{owners} || []};
        my $pattern = $line->{pattern};
        $patterns{$pattern}++ if $pattern;
    }

    my $patterns = [sort keys %patterns];
    $self->{patterns} = $patterns if !$owner;

    return $patterns;
}

=method aliases

    \%aliases = $codeowners->aliases;

Get a hashref of all aliases defined.

=cut

sub aliases {
    my $self = shift;

    return $self->{aliases} if $self->{aliases};

    my %aliases;
    for my $line (@{$self->_lines}) {
        next if !defined $line->{alias};
        $aliases{$line->{alias}} = [@{$line->{owners}}];
    }

    return $self->{aliases} = \%aliases;
}

=method projects

    \@projects = $codeowners->projects;

Get an arrayref of all projects defined.

=cut

sub projects {
    my $self  = shift;

    return $self->{projects} if $self->{projects};

    my %projects;
    for my $line (@{$self->_lines}) {
        my $project = $line->{project};
        $projects{$project}++ if $project;
    }

    my $projects = [sort keys %projects];
    $self->{projects} = $projects;

    return $projects;
}

=method update_owners

    $codeowners->update_owners($pattern => \@new_owners);

Set a new set of owners for a given pattern. If for some reason the file has
multiple such patterns, they will all be updated.

Nothing happens if the file does not already have at least one such pattern.

=cut

sub update_owners {
    my $self    = shift;
    my $pattern = shift;
    my $owners  = shift;
    $pattern && $owners or _usage(q{$codeowners->update_owners($pattern => \@owners)});

    $owners = [$owners] if ref($owners) ne 'ARRAY';

    $self->_clear;

    my $count = 0;

    for my $line (@{$self->_lines}) {
        next if !$line->{pattern};
        next if $pattern ne $line->{pattern};
        $line->{owners} = [@$owners];
        ++$count;
    }

    return $count;
}

=method update_owners_by_project

    $codeowners->update_owners_by_project($project => \@new_owners);

Set a new set of owners for all patterns under the given project.

Nothing happens if the file does not have a project with the given name.

=cut

sub update_owners_by_project {
    my $self    = shift;
    my $project = shift;
    my $owners  = shift;
    $project && $owners or _usage(q{$codeowners->update_owners_by_project($project => \@owners)});

    $owners = [$owners] if ref($owners) ne 'ARRAY';

    $self->_clear;

    my $count = 0;

    for my $line (@{$self->_lines}) {
        next if !$line->{project} || !$line->{owners};
        next if $project ne $line->{project};
        $line->{owners} = [@$owners];
        ++$count;
    }

    return $count;
}

=method rename_owner

    $codeowners->rename_owner($old_name => $new_name);

Rename an owner.

Nothing happens if the file does not have an owner with the old name.

=cut

sub rename_owner {
    my $self        = shift;
    my $old_owner   = shift;
    my $new_owner   = shift;
    $old_owner && $new_owner or _usage(q{$codeowners->rename_owner($owner => $new_owner)});

    $self->_clear;

    my $count = 0;

    for my $line (@{$self->_lines}) {
        next if !exists $line->{owners};
        for (my $i = 0; $i < @{$line->{owners}}; ++$i) {
            next if $line->{owners}[$i] ne $old_owner;
            $line->{owners}[$i] = $new_owner;
            ++$count;
        }
    }

    return $count;
}

=method rename_project

    $codeowners->rename_project($old_name => $new_name);

Rename a project.

Nothing happens if the file does not have a project with the old name.

=cut

sub rename_project {
    my $self        = shift;
    my $old_project = shift;
    my $new_project = shift;
    $old_project && $new_project or _usage(q{$codeowners->rename_project($project => $new_project)});

    $self->_clear;

    my $count = 0;

    for my $line (@{$self->_lines}) {
        next if !exists $line->{project} || $old_project ne $line->{project};
        $line->{project} = $new_project;
        $line->{comment} = " Project: $new_project" if exists $line->{comment};
        ++$count;
    }

    return $count;
}

=method append

    $codeowners->append(comment => $str);
    $codeowners->append(pattern => $pattern, owners => \@owners);
    $codeowners->append();     # blank line

Append a new line.

=cut

sub append {
    my $self = shift;
    $self->_clear;
    push @{$self->_lines}, (@_ ? {@_} : undef);
}

=method prepend

    $codeowners->prepend(comment => $str);
    $codeowners->prepend(pattern => $pattern, owners => \@owners);
    $codeowners->prepend();    # blank line

Prepend a new line.

=cut

sub prepend {
    my $self = shift;
    $self->_clear;
    unshift @{$self->_lines}, (@_ ? {@_} : undef);
}

=method unowned

    \@filepaths = $codeowners->unowned;

Get the list of filepaths in the "unowned" section.

This parser supports an "extension" to the F<CODEOWNERS> file format which
lists unowned files at the end of the file. This list can be useful to have in
order to figure out what files we know are unowned versus what files we don't
know are unowned.

=cut

sub unowned {
    my $self = shift;
    [sort keys %{$self->{unowned} || {}}];
}

=method add_unowned

    $codeowners->add_unowned($filepath, ...);

Add one or more filepaths to the "unowned" list.

This method does not check to make sure the filepath(s) actually do not match
any patterns in the file, so you might want to call L</match> first.

See L</unowned> for an explanation.

=cut

sub add_unowned {
    my $self = shift;
    $self->_unowned->{$_}++ for @_;
}

=method remove_unowned

    $codeowners->remove_unowned($filepath, ...);

Remove one or more filepaths from the "unowned" list.

Silently ignores filepaths that are already not listed.

See L</unowned> for an explanation.

=cut

sub remove_unowned {
    my $self = shift;
    delete $self->_unowned->{$_} for @_;
}


=method is_unowned

    $bool = $codeowners->is_unowned($filepath);

Test whether a filepath is in the "unowned" list.

See L</unowned> for an explanation.

=cut

sub is_unowned {
    my $self     = shift;
    my $filepath = shift;
    $self->_unowned->{$filepath};
}

=method clear_unowned

    $codeowners->clear_unowned;

Remove all filepaths from the "unowned" list.

See L</unowned> for an explanation.

=cut

sub clear_unowned {
    my $self = shift;
    $self->{unowned} = {};
}

sub _lines   { shift->{lines}   ||= [] }
sub _unowned { shift->{unowned} ||= {} }

sub _clear {
    my $self = shift;
    delete $self->{match_lines};
    delete $self->{owners};
    delete $self->{patterns};
    delete $self->{aliases};
    delete $self->{projects};
}

=head1 DESCRIPTION

This module parses and generates F<CODEOWNERS> files.

See L<CODEOWNERS syntax|https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-syntax>.

=cut

1;


lib/File/Codeowners/Util.pm
package File::Codeowners::Util;
# ABSTRACT: Grab bag of utility subs for Codeowners modules


use warnings;
use strict;

use Encode qw(decode);
use Exporter qw(import);
use Path::Tiny;

our @EXPORT_OK = qw(
    find_codeowners_in_directory
    find_nearest_codeowners
    git_ls_files
    git_toplevel
    run_command
    run_git
);

our $VERSION = '0.52'; # VERSION


sub find_nearest_codeowners {
    my $path = path(shift || '.')->absolute;

    while (!$path->is_rootdir) {
        my $filepath = find_codeowners_in_directory($path);
        return $filepath if $filepath;
        $path = $path->parent;
    }
}


sub find_codeowners_in_directory {
    my $path = path(shift) or die;

    my @tries = (
        [qw(CODEOWNERS)],
        [qw(docs CODEOWNERS)],
        [qw(.bitbucket CODEOWNERS)],
        [qw(.github CODEOWNERS)],
        [qw(.gitlab CODEOWNERS)],
    );

    for my $parts (@tries) {
        my $try = $path->child(@$parts);
        return $try if $try->is_file;
    }
}


sub run_command {
    my $filter;
    $filter = pop if ref($_[-1]) eq 'CODE';

    print STDERR "# @_\n" if $ENV{FILE_CODEOWNERS_DEBUG};

    my ($child_in, $child_out);
    require IPC::Open2;
    my $pid = IPC::Open2::open2($child_out, $child_in, @_);
    close($child_in);

    binmode($child_out, ':encoding(UTF-8)');

    my $proc = File::Codeowners::Util::Process->new(
        pid     => $pid,
        fh      => $child_out,
        filter  => $filter,
    );

    return wantarray ? ($proc, @{$proc->all}) : $proc;
}


sub run_git {
    return run_command('git', @_);
}


sub git_toplevel {
    my $dir = shift || '.';

    my ($proc, $path) = run_git('-C', $dir, qw{rev-parse --show-toplevel});

    return if $proc->wait != 0 || !$path;
    return path($path);
}


sub git_ls_files {
    my $dir = shift || '.';
    return run_git('-C', $dir, 'ls-files', @_, \&_unescape_git_filepath);
}

# Depending on git's "core.quotepath" config, non-ASCII chars may be
# escaped (identified by surrounding dquotes), so try to unescape.
sub _unescape_git_filepath {
    return $_ if $_ !~ /^"(.+)"$/;
    return decode('UTF-8', _unbackslash($1));
}

# The unbacklash code is from String::Escape (thanks EVO), with changes:
# - Handle \a, \b, \f and \v (thanks Berk Akinci)
my %unbackslash;
sub _unbackslash {
    my $str = shift;
    # Earlier definitions are preferred to later ones, thus we output \n not \x0d
    %unbackslash = (
        ( map { $_ => $_ } ( '\\', '"', '$', '@' ) ),
        ( 'r' => "\r", 'n' => "\n", 't' => "\t" ),
        ( map { 'x' . unpack('H2', chr($_)) => chr($_) } (0..255) ),
        ( map { sprintf('%03o', $_) => chr($_) } (0..255) ),
        ( 'a' => "\x07", 'b' => "\x08", 'f' => "\x0c", 'v' => "\x0b" ),
    ) if !%unbackslash;
    $str =~ s/ (\A|\G|[^\\]) \\ ( [0-7]{3} | x[\da-fA-F]{2} | . ) / $1 . $unbackslash{lc($2)} /gsxe;
    return $str;
}

{
    package File::Codeowners::Util::Process;

    sub new {
        my $class = shift;
        return bless {@_}, $class;
    }

    sub next {
        my $self = shift;
        my $line = readline($self->{fh});
        if (defined $line) {
            chomp $line;
            if (my $filter = $self->{filter}) {
                local $_ = $line;
                $line = $filter->($line);
            }
        }
        $line;
    }

    sub all {
        my $self = shift;
        chomp(my @lines = readline($self->{fh}));
        if (my $filter = $self->{filter}) {
            $_ = $filter->($_) for @lines;
        }
        \@lines;
    }

    sub wait {
        my $self = shift;
        my $pid  = $self->{pid} or return;
        if (my $fh = $self->{fh}) {
            close($fh);
            delete $self->{fh};
        }
        waitpid($pid, 0);
        my $status = $?;
        print STDERR "# -> status $status\n" if $ENV{FILE_CODEOWNERS_DEBUG};
        delete $self->{pid};
        return $status;
    }

    sub DESTROY {
        my ($self, $global_destruction) = @_;
        return if $global_destruction;
        $self->wait;
    }
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

File::Codeowners::Util - Grab bag of utility subs for Codeowners modules

=head1 VERSION

version 0.52

=head1 DESCRIPTION

B<DO NOT USE> except in L<File::Codeowners> and related modules.

=head1 FUNCTIONS

=head2 find_nearest_codeowners

    $filepath = find_nearest_codeowners($dirpath);

Find the F<CODEOWNERS> file in the current working directory, or search in the
parent directory recursively until a F<CODEOWNERS> file is found.

Returns C<undef> if no F<CODEOWNERS> is found.

=head2 find_codeowners_in_directory

    $filepath = find_codeowners_in_directory($dirpath);

Find the F<CODEOWNERS> file in a given directory. No recursive searching is done.

Returns the first of (or undef if none found):

=over 4

=item *

F<CODEOWNERS>

=item *

F<docs/CODEOWNERS>

=item *

F<.bitbucket/CODEOWNERS>

=item *

F<.github/CODEOWNERS>

=item *

F<.gitlab/CODEOWNERS>

=back

=head2 run_command

    $proc = run_command(@cmd);
    $proc = run_command(@cmd, $filter);

    while (defined (my $line = $proc->next)) { ...  }
    $status = $proc->wait;

Run a command and read its output, optionally filtering each line.

=head2 run_git

    $proc = run_git(@args);
    $proc = run_git(@args, $filter);

Run C<git>.

=head2 git_toplevel

    $path = git_toplevel();
    $path = git_toplevel($directory);

Get a L<Path::Tiny> of the toplevel directory for a git repository, or C<undef> of there is none.

=head2 git_ls_files

    $proc = git_ls_files();
    $proc = git_ls_files($directory);

Run C<git ls-files> while handling path quoting.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
L<https://github.com/chazmcgarvey/File-Codeowners/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Charles McGarvey <chazmcgarvey@brokenzipper.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2021 by Charles McGarvey.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.014.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Read and write CODEOWNERS files",
  "AUTHOR" => "Charles McGarvey <chazmcgarvey\@brokenzipper.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "File-Codeowners",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "File::Codeowners",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Encode" => 0,
    "Exporter" => 0,
    "IPC::Open2" => 0,
    "Path::Tiny" => "0.089",
    "Scalar::Util" => 0,
    "Text::Gitignore" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.52",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Encode" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "IPC::Open2" => 0,
  "IPC::Open3" => 0,
  "Path::Tiny" => "0.089",
  "Scalar::Util" => 0,
  "Test::More" => 0,
  "Text::Gitignore" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
